### PR TITLE
Flush logical messages with snapshots and replication origin

### PR DESCRIPTION
**Pull request by knizhnik**
## Problem

See https://neondb.slack.com/archives/C04DGM6SMTM/p1708363190710839

## Summary of changes

Flush logical message with snapshot and origin state

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Comment by knizhnik** [comment]
> This contains lots of unrelated changes; I'm not sure we should have those in this PR specifically. But, apart from that, the changes in `src/backend/replication/logical/snapbuild.c` and `src/backend/replication/logical/origin.c` should indeed do their job (or at least reduce the prevalence of the issue)

Sorry, I have not committed unrelated changed.
If you see my commit 5baf0df00dafad26d8be98b9cef7d3bca80d6c4f then you can check that it fix only few lines of code (adding XLogFlush).

I think that the reasons of so much changes reported by GIT are caused by the fact that there were many recent commits in postgres repository but version of Postgres in neon repository was not bumped.

**Comment by save-buffer** [comment]
I approved the PRs in the postgres repos, but those should get rebased before being merged. The commit hashes here ned to be updated too, I'll approve once the postgres repo PRs are merged

**Comment by knizhnik** [comment]
@save-buffer can you please approve this PR to bump Postgres version?

**Review by save-buffer — APPROVED** [review]

**Pull request merged**
